tests.py

```diff
@@ -1,129 +1,523 @@
-import contextlib
 import pathlib
 import platform
 import shutil
 import subprocess
 import sys
 import textwrap
+import re
+import threading
+import queue
+import time
+import email
+from http import client
 
 import pytest
 
 import sansio_lsp_client as lsp
 
 
-@contextlib.contextmanager
-def run_langserver(project_root, command):
+LOG_IN = False  # log data received from server to stdout
+LOG_OUT = False  # - sent - to stdout
+
+
+METHOD_DID_OPEN = "didOpen"
+METHOD_DID_CLOSE = "didClose"
+METHOD_DID_SAVE = "didSave"
+METHOD_DID_CHANGE = "didChange"
+
+METHOD_COMPLETION = "completion"
+METHOD_HOVER = "hover"
+METHOD_SIG_HELP = "signatureHelp"
+METHOD_DEFINITION = "definition"
+METHOD_REFERENCES = "references"
+METHOD_IMPLEMENTATION = "implementation"
+METHOD_DECLARATION = "declaration"
+METHOD_TYPEDEF = "typeDefinition"
+METHOD_DOC_SYMBOLS = "documentSymbol"
+METHOD_FORMAT_DOC = "formatting"
+METHOD_FORMAT_SEL = "rangeFormatting"
+
+METHOD_WS_FOLDERS = "workspace/workspaceFolders"
+
+RESPONSE_TYPES = {
+    METHOD_COMPLETION: lsp.Completion,
+    METHOD_HOVER: lsp.Hover,
+    METHOD_SIG_HELP: lsp.SignatureHelp,
+    METHOD_DEFINITION: lsp.Definition,
+    METHOD_REFERENCES: lsp.References,
+    METHOD_IMPLEMENTATION: lsp.Implementation,
+    METHOD_DECLARATION: lsp.Declaration,
+    METHOD_TYPEDEF: lsp.TypeDefinition,
+    METHOD_DOC_SYMBOLS: lsp.MDocumentSymbols,
+    METHOD_FORMAT_DOC: lsp.DocumentFormatting,
+    METHOD_FORMAT_SEL: lsp.DocumentFormatting,
+}
+
+
+_MAXLINE = 65536
+_MAXHEADERS = 100
+
+# from https://github.com/python/cpython/blob/3.9/Lib/http/client.py
+#  modified to keep header_bytes
+def parse_headers(fp, _class=client.HTTPMessage):
```

Comment:
There's header parsing code in sansio_lsp_client/io_handler.py. Why is this needed?

Reply:
`_parse_one_message()` - expects 'header + body' bytearray, I need to read data from the server (stdout) first - that's what this function does

Reply:
You can pass incomplete messages to sansio-lsp-client. It will handle them correctly.

Reply:
I guess I should clarify this a bit. `_parse_one_message()` expects header + body, but:
- it doesn't consume the bytearray if it doesn't contain the full body or headers
- if there's something after the body, it leaves it to the bytearray

For this reason, it's fine to call it repeatedly until there is no longer full headers and body left (i.e. it returns None). This is exactly what `_parse_messages()` does, and you should just use `_parse_messages()` (or even better, `Client.recv()`, which does that for you).

Reply:
Ok... I played with it and it seems like the only way to use `_parse_messages()` (`recv()`) - to read one byte at a time from server, which doesn't seem like a good way to go...

Basically to do a proper job I need to get body size while reading `stdout` -- a function to parse headers only

Reply:
One byte at a time is fine for tests. It is more difficult than that in general: https://github.com/Akuli/porcupine/blob/4d211dfee6ceb9034e7ca0660f6e243c5152220b/porcupine/plugins/langserver.py#L42-L97